pkg/controller/drain/drain_controller.go

```diff
@@ -0,0 +1,425 @@
+	}
+
+	<-stopCh
+}
+
+func (ctrl *Controller) handleNodeEvent(node interface{}) {
+	n := node.(*corev1.Node)
+	glog.V(4).Infof("Updating Node %s", n.Name)
+	ctrl.enqueueNode(n)
+}
+
+func (ctrl *Controller) enqueue(node *corev1.Node) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+		return
+	}
+
+	ctrl.queue.Add(key)
+}
+
+func (ctrl *Controller) enqueueRateLimited(node *corev1.Node) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+		return
+	}
+
+	ctrl.queue.AddRateLimited(key)
+}
+
+// enqueueAfter will enqueue a pool after the provided amount of time.
+func (ctrl *Controller) enqueueAfter(node *corev1.Node, after time.Duration) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+		return
+	}
+
+	ctrl.queue.AddAfter(key, after)
+}
+
+// enqueueDefault calls a default enqueue function
+func (ctrl *Controller) enqueueDefault(node *corev1.Node) {
+	ctrl.enqueueAfter(node, updateDelay)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (ctrl *Controller) worker() {
+	for ctrl.processNextWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextWorkItem() bool {
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	err := ctrl.syncHandler(key.(string))
+	ctrl.handleErr(err, key)
+
+	return true
+}
+
+func (ctrl *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.queue.Forget(key)
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing node %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(2).Infof("Dropping node %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+func (ctrl *Controller) syncNode(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing node %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing node %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// First check if a drain request is needed for this node
+	node, err := ctrl.nodeLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("node %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	desiredState := node.Annotations[daemonconsts.DesiredDrainerAnnotationKey]
+	if desiredState == node.Annotations[daemonconsts.LastAppliedDrainerAnnotationKey] {
```

Comment:
One thing I found confusing here is the case of "newly joined node" (initial cluster install or scaled up).  What I expected to see is code around here that says "no drain requested".

But it looks like we end dropping down into...the "uncordon" flow?

Reply:
Ah yeah there should be a nil check here, which hypershift has. I will add that on the next iteration